### PR TITLE
New paradigm for bundling plugin dependencies and `files` bundles

### LINKED ISSUE #3355: New paradigm for bundling plugin dependencies and `files` bundles
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3433

Originally created by @aaronsteers on 2022-04-23 23:11:08

---

Today:

```yml
transformers:
  - name: dbt
    namespace: dbt
    docs: https://docs.meltano.com/guide/transformation
    repo: https://github.com/dbt-labs/dbt-core
    pip_url: dbt-core~=1.0.0 dbt-postgres~=1.0.0 dbt-redshift~=1.0.0 dbt-snowflake~=1.0.0 dbt-bigquery~=1.0.0
    settings:
    # ...
orchestrators:
  - name: airflow
    namespace: airflow
    docs: https://docs.meltano.com/guide/orchestration
    repo: https://github.com/apache/airflow
    pip_url: 'apache-airflow==2.1.2 --constraint https://raw.githubusercontent.com/apache/airflow/constraints-2.1.2/constraints-${MELTANO__PYTHON_VERSION}.txt'
    settings:
    # ...
# ...
files:
  - name: airflow
    namespace: airflow # << auto-installed with `dbt` due to matching `namespace`
    repo: https://gitlab.com/meltano/files-airflow
    pip_url: git+https://gitlab.com/meltano/files-airflow.git
    update:
      orchestrate/dags/meltano.py: true
  - name: dbt
    namespace: dbt # << auto-installed with `dbt` due to matching `namespace`
    repo: https://gitlab.com/meltano/files-dbt
    pip_url: git+https://gitlab.com/meltano/files-dbt.git@config-version-2
```

## The challenge

The number of collisions on namespace will potentially skyrocket one we complete [#3283](https://gitlab.com/meltano/meltano/-/issues/3283). We can no longer reasonably scan all the plugins in the hub or in discovery.yml to identify matching namespaces and then have confidence that the matching plugins should be installed together. Instead, we need a more explicit mapping of files resources to the plugins that need those resources.

## Proposed refactor:

- The `namespace` field can be removed (optionally auto-calculated internally per [#3374](https://gitlab.com/meltano/meltano/-/issues/3374)).
- Files resources are now defined inline with the plugins themselves.

```yml
transformers:
  - name: dbt
    docs: https://docs.meltano.com/guide/transformation
    repo: https://github.com/dbt-labs/dbt-core
    pip_url: dbt-core~=1.0.0 dbt-postgres~=1.0.0 dbt-redshift~=1.0.0 dbt-snowflake~=1.0.0 dbt-bigquery~=1.0.0
    bundle: # or `depends_on:`
      files:
      - name: files-dbt  # Name might be optional, since the plugin has a name in its API definition.
        # This is the API endpoint
        ref: https://hub.meltano.com/meltano/api/v1/plugins/extractors/tap-zendesk--twilio-labs
orchestrators:
  - name: airflow
    docs: https://docs.meltano.com/guide/orchestration
    repo: https://github.com/apache/airflow
    pip_url: 'apache-airflow==2.1.2 --constraint https://raw.githubusercontent.com/apache/airflow/constraints-2.1.2/constraints-${MELTANO__PYTHON_VERSION}.txt'
    bundle:
      files:
      - name: airflow-files
        variant: meltanolabs
    settings:
    # ...
```

The files needed for the repo would then be self-contained within the plugin definition and are self-sufficient for installing the plugin, without needing external plugin to be installed.

## Moving to `copier`

Our current paradigm does not facilitate supporting changes over time to the files (the "day 2" problem). A related proposal would be to [#3432](https://gitlab.com/meltano/meltano/-/issues/3432), but that can be delivered separately from this change.

## Standalone `files` plugins

We can still support standalone `files` plugins for use cases where it still makes sense to do so apart from auto-installed resources for a `utility`, `transformer`, or `orchestrator` plugin.

However, we would not post separate definitions to the hub for those `files` plugins which are best combined with their particular plugin.

### PR DESCRIPTION
**Pull request by edgarrmondragon**
Potential blockers:

- https://github.com/meltano/meltano/pull/5957/

Closes #3355

### CODE REVIEW

**Comment by edgarrmondragon** [comment]
Probably blocked by https://github.com/meltano/meltano/pull/5957

**Comment by aaronsteers** [comment]
Thanks for kicking this off, @edgarrmondragon.

Example with `requires` entry here: https://github.com/meltano/hub/pull/505

File: tests/meltano/cli/test_add.py [hunk 1]
```diff
@@ -23,18 +23,19 @@ def patch_hub(self, meltano_hub_service: MeltanoHubService):
...
-            (
+            pytest.param(
                 PluginType.TRANSFORMS,
                 "tap-carbon-intensity",
                 "meltano",
                 [
                     PluginRef(PluginType.TRANSFORMERS, "dbt"),
                     PluginRef(PluginType.FILES, "dbt"),
                 ],
+                marks=[pytest.mark.xfail],
```
Comment (edgarrmondragon) [inline tests/meltano/cli/test_add.py:38]: Unsure what to do with this. Maybe wait for https://github.com/meltano/meltano/pull/5957 to merge.

File: tests/meltano/cli/test_add.py [hunk 2]
```diff
@@ -23,18 +23,19 @@ def patch_hub(self, meltano_hub_service: MeltanoHubService):
...
             ),
```
Comment (aaronsteers) [inline tests/meltano/cli/test_add.py:39]:
```suggestion
            pytest.param(
                PluginType.TRANSFORMS,
                "tap-carbon-intensity",
                "meltano",
                [
                    PluginRef(PluginType.TRANSFORMERS, "dbt"),
                    PluginRef(PluginType.FILES, "dbt"),
                ],
                marks=[pytest.mark.xfail], # Will be removed in: https://github.com/meltano/meltano/pull/5957
            ),
```

File: tests/meltano/cli/test_add.py (hunk 1 above)
Comment (aaronsteers) [reply tests/meltano/cli/test_add.py:38]: XFail or delete - I think either is fine.

Suggestion here links to the related info as explanation: https://github.com/meltano/meltano/pull/5963/files#r888434730

Either way, that issue will have a marge conflict to resolve - but should be manageable.

**Review by aaronsteers — APPROVED** [review]
@edgarrmondragon - This looks good to my eyes. If possible to merge and move forward without blocking on https://github.com/meltano/meltano/issues/3355, that would be my preference.

(`pytest_fast` has passed successfully but other tests are still running, fwiw.)

File: tests/meltano/cli/test_add.py (hunk 1 above)
Comment (edgarrmondragon) [reply tests/meltano/cli/test_add.py]: Alright, happy merge this and help resolve conflicts in #5957

**Comment by edgarrmondragon** [comment]
@aaronsteers 

> Is it helpful if I test locally?

If you can spare some time that'd be good, but I'm also happy to merge now and let folks test master in case there's any corner case not considered here

File: src/meltano/cli/add.py [hunk 3]
```diff
@@ -112,10 +116,13 @@ def add(
...
+    required_plugins = add_required_plugins(
+        project,
+        plugins,
+        add_service=add_service,
+        plugin_types=related_plugin_types,
     )
-    plugins.extend(related_plugins)
+    plugins.extend(required_plugins)
 
     # We will install the plugins in reverse order, since dependencies
```
Comment (kgpayne) [inline src/meltano/cli/add.py:127]: @edgarrmondragon does the new `required:` functionality rely on this reverse order install? It was removed by #5957 as a redundant requirement of the previous `add_related` functionality, but I see here it is tested working with the new `required_plugins` extended list 🤔

Comment (edgarrmondragon) [reply src/meltano/cli/add.py:127]: @kgpayne Nope, it shouldn't rely on it. The diff is a bit weird cause this is definitely not present in `main`
